### PR TITLE
support knowhere 2.0

### PR DESCRIPTION
**Pull request by presburger**
issues: #21858
Signed-off-by: Yusheng.Ma <Yusheng.Ma@zilliz.com>

### CODE REVIEW

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

**Comment by Presburger** [comment]
rerun ut

File: internal/core/src/index/VectorIndex.h [hunk 1]
```diff
@@ -81,5 +81,5 @@ class VectorIndex : public IndexBase {
     int64_t dim_;
 };
 
-using VectorIndexPtr = std::unique_ptr<VectorIndex>;
+using VectorIndexPtr = std::unique_ptr<knowhere::Index<knowhere::IndexNode>>;
```
Comment (xige-16) [inline internal/core/src/index/VectorIndex.h:84]: knowhere::IndexNode seems to be a type of knowhre that is not exposed and should not be used directly in milvus

File: internal/core/src/query/visitors/VerifyPlanNodeVisitor.cpp [hunk 2]
```diff
@@ -66,38 +64,10 @@ InferBinaryIndexType(const Json& search_params) {
 
 void
 VerifyPlanNodeVisitor::visit(FloatVectorANNS& node) {
```
Comment (xige-16) [inline internal/core/src/query/visitors/VerifyPlanNodeVisitor.cpp:66]: All impl of  VerifyPlanNodeVisitor seems can be removed.

File: internal/core/src/index/VectorIndex.h (hunk 1 above)
Comment (xige-16) [reply internal/core/src/index/VectorIndex.h:84]: And this change breaks the inheritance relationship between VectorIndexPtr and indexbase, VectorDiskannIndex, vectorMemIndex

File: internal/core/unittest/test_c_api.cpp [hunk 3]
```diff
@@ -1400,8 +1396,8 @@ TEST(CApiTest, LoadIndexInfo) {
 
     auto N = 1024 * 10;
     auto [raw_data, timestamps, uids] = generate_data(N);
-    auto indexing = std::make_shared<knowhere::IVFPQ>();
-    auto conf = knowhere::Config{{knowhere::meta::METRIC_TYPE, knowhere::metric::L2},
+    auto indexing = knowhere::IndexFactory::Instance().Create("IVFPQ");
```
Comment (cydrain) [inline internal/core/unittest/test_c_api.cpp:1399]: should use MACRO instead of raw string

**Comment by Presburger** [comment]
rerun ut

**Pull request merged**